---
title: "Use Case - Annotation Hub and GRanges, Part 1"
author: "Sasha Ajay Malkani"
date: "2025-10-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Installing the AnnotationHub package

```{r Installing the AnnotationHub package}
# BiocManager::install("AnnotationHub")

```

## 2. Load AnnotationHub

```{r Load AnnotationHub}
library(AnnotationHub)

```

## 3. Instantiate an Annotation Hub

```{r Instantiate an Annotation Hub}
ahub = AnnotationHub()
ahub

```

## 4. Subset the Annotation Hub to be specific to Human data

```{r Subset the Annotation Hub to be specific to Human data}
ahub = subset(ahub, species == "Homo sapiens")

```

## 5. Search for data under two search terms using query function

```{r Search for data under two search terms using query function}
qhs = query(ahub, c("H3K4me3", "Gm12878"))
qhs

```

## 6. Get two genome ranges the second element and the fourth element

```{r Get two genome ranges the second element and the fourth element}
# 2nd element Rep1, broadPeak = gr1
# 4th element Rep1, narrowPeak = gr2
gr1 = qhs[[2]]
gr2 = qhs[[4]]

```

## 7. 2nd Element

```{r 2nd Element}
gr1

```

## 8. Summary of 2nd element (broad peak)

```{r Summary of 2nd element (broad peak)}
summary(width(gr1))

```

## 9. Summary of 4th element (narrow peak)

```{r Summary of 4th element (narrow peak)}
summary(width(gr2))

```

## 10. Create a table of all the widths for the narrow peak

```{r Create a table of all the widths for the narrow peak}
table(width(gr2))

```

## 11. Rename the GRanges (gr2) to peaks

```{r Rename the GRanges (gr2) to peaks}
peaks = gr2

```

## 12. Re-list the genomes

```{r Re-list the genomes}
qhs

```

## 13. Display the Rep1, narrow peak genome

```{r Display the Rep1, narrow peak genome}
qhs[4]

```

## 14. Performing a query on the RefSeq database

```{r Performing a query on the RefSeq database}
qhs = query(ahub, "RefSeq")
qhs

```

## 15. Looked at the genome (clarify why we have 8 data sets)

```{r Looked at the genome}
qhs$genome

```

## 16. Download the first data set from ahub the hg19 gene

```{r Download the first data set from ahub the hg19 gene}
genes = qhs[[1]]
genes

```

## 17. Use a table to identify how often a gene occurs

Nesting the table function within a table function, to see how often a single or two transcripts occur

```{r Use a table to identify how often a gene occurs}
table(table(genes$name))

```

## 18. Call promoters on your data

```{r 18. Call promoters on your data}
prom = promoters(genes)

```

## 19. Define the width of a promoter

```{r 19. Define the width of a promoter}
table(width(prom))

```

## 20. Look at the arguments for promoters

```{r 20. Look at the arguments for promoters}
args(promoters)
function (x, upstream = 2000, downstream = 200, ...) 
  NULL

```


